### PR TITLE
refactor(tokens): integer math for token deposit/withdraw

### PR DESCRIPTION
**Pull request by raul-oliveira**
## Motivation

Mirrors hathor-core [PR #1711](https://github.com/HathorNetwork/hathor-core/pull/1711) on the client side.

HTR deposit/withdraw for deposit-based tokens applied the deposit percentage with float math, which loses precision above the JS safe-integer boundary (2^53) and produces wrong values for large amounts still within `MAX_OUTPUT_VALUE` (2^63). This PR uses integer math instead, with the percentage as a parts-per-billion fraction (`10^7 / 10^9 = 1%`) matching hathor-core. No behavior change for valid configurations.

### Acceptance Criteria

- Deposit/withdraw calculations are exact for amounts above 2^53 (within `MAX_OUTPUT_VALUE`).
- No behavior change for valid existing configurations.
- Backward compatible with fullnodes that don't expose the new `/version` fields.

## Follow-up (not in this PR)

End-to-end integration test against a fullnode image with #1711 is deferred to #1117: published images with #1711 are 85 commits ahead of the pinned baseline and need a privnet config migration first.

### Security Checklist

- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

## Release Notes

* **Bug Fixes**
  * Corrected token deposit/withdraw fee calculations to use exact bigint fraction math, avoiding floating-point precision issues and ensuring correct deposit ceiling vs withdrawal flooring.
* **New Features**
  * Added integer-precision token deposit percentage support via numerator/denominator with legacy fallback.
  * Exposed wallet facade methods to compute deposit/withdraw amounts using the stored fraction.
  * Extended version data and API schemas to optionally provide the fraction fields (legacy field is deprecated).
* **Tests**
  * Added edge-case coverage for negative inputs and large bigint rounding behavior.

### CODE REVIEW

File: src/storage/storage.ts [hunk 1]
```diff
@@ -205,6 +209,37 @@ export class Storage implements IStorage {
...
+   * that predate these fields), it falls back to deriving the numerator from the deprecated
+   * float percentage, keeping the standard 1e9 denominator, and ultimately to the defaults.
+   *
+   * @returns {{ numerator: bigint; denominator: bigint }}
+   */
+  getTokenDepositPercentageFraction(): { numerator: bigint; denominator: bigint } {
+    const numerator = this.version?.token_deposit_percentage_numerator;
+    const denominator = this.version?.token_deposit_percentage_denominator;
+    if (numerator != null && denominator != null) {
+      return { numerator: BigInt(numerator), denominator: BigInt(denominator) };
```
Comment (tuliomir) [inline src/storage/storage.ts:241]: chore: we should have all flows this function tested in isolation by a unit test suite

File: src/utils/bigint.ts [hunk 2]
```diff
@@ -66,6 +66,18 @@ export const JSONBigInt = {
...
 };
 
+/**
+ * Ceil division using integer math for non-negative operands, equivalent to `ceil(a / b)`.
+ *
+ * Avoids the precision loss of `Math.ceil(a / b)` with floats for large values.
+ */
+export function ceilDiv(a: bigint, b: bigint): bigint {
+  if (a < 0n || b <= 0n) {
+    throw new Error('ceilDiv expects a non-negative dividend and a positive divisor');
```
Comment (tuliomir) [inline src/utils/bigint.ts:76]: chore: we should unit test this error handler

Comment (raul-oliveira) [reply src/utils/bigint.ts:76]: Added a `ceilDiv` suite covering the guard (negative dividend, zero/negative divisor) plus zero-dividend, exact, and round-up cases.

File: src/storage/storage.ts (hunk 1 above)
Comment (raul-oliveira) [reply src/storage/storage.ts:241]: Added a `getTokenDepositPercentageFraction` suite covering every branch: fields preferred when both present, derivation from the float percentage (including when only one field is set), and the default fallback for unset/deposit-less versions.

File: src/template/transaction/executor.ts [hunk 3]
```diff
@@ -626,8 +626,10 @@ export async function execCompleteTxInstruction(
     let deposit = 0n;
 
     if (ctx.tokenVersion === TokenVersion.DEPOSIT) {
-      const amount = ctx.balance.createdTokenBalance!.tokens;
-      deposit = interpreter.getHTRDeposit(amount);
+      // createdTokenBalance is negative, so we use its absolute value.
```
Comment (tuliomir) [inline src/template/transaction/executor.ts:629]: question(non-blocking): Should we really mutate the input to match our validation rules? Wouldn't it be better to throw instead?

**Review by tuliomir — DISMISSED** [review]

File: src/wallet/api/schemas/walletApi.ts [hunk 4]
```diff
@@ -237,6 +237,8 @@ export const fullNodeVersionDataSchema = z
     minTxWeightCoefficient: z.number(),
     minTxWeightK: z.number(),
     tokenDepositPercentage: z.number(),
```
Comment (pedroferreira1) [inline src/wallet/api/schemas/walletApi.ts]: This is deprecated and should be removed in the next node versions. We should prepare the wallet for it already, otherwise we would need a new update to stop requiring this field

File: src/wallet/types.ts [hunk 5]
```diff
@@ -642,7 +642,12 @@ export interface FullNodeVersionData {
   minTxWeight: number;
   minTxWeightCoefficient: number;
   minTxWeightK: number;
+  /** @deprecated Use the numerator/denominator fraction below, which keeps integer precision. */
   tokenDepositPercentage: number;
```
Comment (pedroferreira1) [inline src/wallet/types.ts]: Same comment as in the schema

File: src/types.ts [hunk 6]
```diff
@@ -585,7 +585,12 @@ export interface ApiVersion {
   min_tx_weight: number;
   min_tx_weight_coefficient: number;
   min_tx_weight_k: number;
+  /** @deprecated Use the numerator/denominator fraction below, which keeps integer precision. */
   token_deposit_percentage: number;
```
Comment (pedroferreira1) [inline src/types.ts]: Same comment as in the others

File: src/utils/tokens.ts [hunk 7]
```diff
@@ -299,25 +301,28 @@ const tokens = {
...
+    depositDenominator: bigint = TOKEN_DEPOSIT_PERCENTAGE_DENOMINATOR
   ): OutputValueType {
-    // This conversion from mintAmount to Number may cause loss of precision for large amounts,
-    // but this is fully equivalent to the reference Python implementation, which does the same.
-    // It'll never be a problem for mainnet as no values can reach the precision boundary, but
-    // it may happen in custom networks.
-    return BigInt(Math.ceil(depositPercent * Number(mintAmount)));
+    // Coerce to bigint so callers passing a number keep working (the API previously used Number()).
+    const amount = BigInt(mintAmount);
+    if (mintAmount < 0) {
```
Comment (pedroferreira1) [inline src/utils/tokens.ts:322]:
```suggestion
    if (mintAmount <= 0) {
```

**Review by pedroferreira1 — COMMENTED** [review]
This is a PR that adds breaking changes to the wallets.

The wallets desktop and mobile (maybe others as well) use the method `getDepositAmount` and `getWithdrawAmount` that changed their parameters.

If that's really the case, we need to merge this carefully and think about the clients updates.

Also, this reflects a bad design decision. I prefer having a method in the wallet facades `wallet.getDepositAmount` and `wallet.getWithdrawAmount` that will get the percentage parameters from `this` correctly and then call the utils method. The interface methods with the clients don't need parameter and will never change, so this breaking change wouldn't happen if it had been designed like this.

Maybe we should add these facade methods in this PR already to improve this. What do you think?

File: src/utils/tokens.ts (hunk 7 above)
Comment (pedroferreira1) [reply src/utils/tokens.ts:322]: Or should it allow 0? If it should, we must just update the error message below

File: src/template/transaction/executor.ts (hunk 3 above)
Comment (raul-oliveira) [reply src/template/transaction/executor.ts:629]: is this case, is needed for the calculations inside the builder. We use positives for inputs and negatives for outputs, so it's a rule that is already derived inside the executor.

File: src/wallet/api/schemas/walletApi.ts (hunk 4 above)
Comment (raul-oliveira) [reply src/wallet/api/schemas/walletApi.ts]: Done — `tokenDepositPercentage` is now `z.number().optional()` + `@deprecated`, so validation survives fullnodes dropping it.

File: src/wallet/types.ts (hunk 5 above)
Comment (raul-oliveira) [reply src/wallet/types.ts]: Same fix on `FullNodeVersionData`: `tokenDepositPercentage?` optional + `@deprecated`.

File: src/types.ts (hunk 6 above)
Comment (raul-oliveira) [reply src/types.ts]: Same fix on `ApiVersion`: `token_deposit_percentage?` optional + `@deprecated`. Readers already fall back to the defaults, so nothing breaks.

File: src/utils/tokens.ts (hunk 7 above)
Comment (raul-oliveira) [reply src/utils/tokens.ts:322]: Now allows 0 (`getDepositAmount(0n) === 0n`); only negatives throw, with the message updated to `mint amount should not be negative`.

**Comment by raul-oliveira** [comment]
@pedroferreira1 agreed. Added `getDepositAmount(mintAmount)` / `getWithdrawAmount(meltAmount)` to both facades (`HathorWallet`, `HathorWalletServiceWallet`) and `IHathorWallet`. They pull the fraction from the wallet's `/version` storage and delegate to the utils, so the client signature takes only the amount and won't break on future percentage-format changes. Sync; they throw `WalletError('Wallet not ready')` instead of falling back to the 1% default. The internal deposit/withdraw sites now route through them.

Follow-up: migrate desktop/mobile off the direct `tokensUtils.getDepositAmount(amount, percent)` calls to the new facade methods.

**Review by pedroferreira1 — APPROVED** [review]

File: __tests__/storage/storage.test.ts [hunk 8]
```diff
@@ -216,6 +218,65 @@ describe('config version', () => {
...
+        token_deposit_percentage_numerator: 3,
+        token_deposit_percentage_denominator: 100,
+      } as ApiVersion);
+      expect(storage.getTokenDepositPercentageFraction()).toEqual({
+        numerator: 3n,
+        denominator: 100n,
+      });
+    });
+
+    it('should derive the fraction from the float percentage when the fields are absent', async () => {
```
Comment (tuliomir) [inline __tests__/storage/storage.test.ts:236]: suggestion(non-blocking): Add a new test to prove that all works after the migration is completed and the `token_deposit_percentage` field is no longer offered by the backend.

**Review by tuliomir — APPROVED** [review]
